### ISSUE TITLE
Understanding the use of pyterrier for my usecase

### ISSUE THREAD

**Issue by sagarkar10** [terrier-org/pyterrier#207]
My main aim is to do a feature extraction (and down the line train models inside pyterrier itself). I have some data in csv with looks something like 
```
query,title,id,price,description
iphone,iphone 12,M101,799,apple iphone with lots of features
mobile,samsung s21,M211,599,not iphone with lot more features
```

So I want to derive my feature sets some are query dependent, some are not. Let's assume the feature set is
```
f_query_doc_1 = some_scoring_func(title, query)
f_query_doc_2 = some_scoring_func(description, query)

f_query_only = len(query)

f_document_only = len(description)
....
```

so as I understand the steps I need to take to achieve this are:

- Index my documents in some supported format with fields: `title, id, price, description`
- Use the `BatchRetrive` function to get the scores using some `wmodel` or something similar
- Use those scores as my features set

This raises following concerns:

- How do I get separate scores for example `some_scoring_func` be `bm25` but that only applies to some fields, and not the entire document.
- How do I get only query features, only document features, and query_document_features.

Thanks for the wonderful module and looking forward to some help getting the best out of it. A small code snippet example regarding my use case will be highly appreciated.

To give a complete scope of my feature requirements, you can refer: https://github.com/ten-blue-links/fxt/blob/master/doc/features.md

**Comment by sagarkar10** [comment]
@cmacdonald

**Comment by cmacdonald** [comment]
Sorry for the delay, will reply later today.

**Comment by sagarkar10** [comment]
@cmacdonald

**Comment by cmacdonald** [comment]
Sorry for the delay.

So there are two ways to do this:

(1) is to use an IterDictIndexer, and set more than one field.  Then, during retrieval, you can FeaturesBatchRetrieve to specify weighting models on the different features.

(2) is to use BatchRetrieve to do your initial retrieval, and then use additional text_scorers to compute your additional features. You use the ** operator to combine those together.

Its worth perusing again https://pyterrier.readthedocs.io/en/latest/ltr.html which I think covers both options.

**Comment by sagarkar10** [comment]
@cmacdonald I appreciate your response, let me rephrase in a simple way.
Lets say I have 2 fields fieldA, fieldB
I want to calculte bm25 of fieldA only with the query.
I can see options to calculate Weights for entire document with query. Can I achive field specific features.

**Comment by cmacdonald** [comment]
There is a weighting model called SingleFieldModel. Have a look at some of the *terrier* documentation examples for that.

Sent from my iPhone

On 1 Aug 2021, at 14:36, Sagar Kar ***@***.***> wrote:

﻿

@cmacdonald<https://github.com/cmacdonald> I appreciate your response, let me rephrase in a simple way.
Lets say I have 2 fields fieldA, fieldB
I want to calculte bm25 of fieldA only with the query.
I can see options to calculate Weights for entire document with query. Can I achive field specific features.

**Comment by cmacdonald** [comment]
E.g. 'SingleFieldModel(BM25,0)' for the first field

**Comment by cmacdonald** [comment]
Hi @sagarkar10 - is there anything we can do to improve the documentation here?

**Comment by sagarkar10** [comment]
@cmacdonald I would like to see a walk through example as I have mentioned previously.

**Comment by sagarkar10** [comment]
@cmacdonald any update?
